### PR TITLE
Cache aware time series

### LINKED ISSUE #5017: Slow cached iter for time series on main
I'm getting surprisingly slow performance on a time series plot on main. It's a single plot with 1M scalars and iter is taking 79+ms when cached. Is that not quite slow for cached values?

Screenshot 2024-02-02 at 13 25 31

Repro by loading https://rerunio.slack.com/files/U02UE9CAWF9/F06GG4YPHQF/1m_series.rrd on main

### PR DESCRIPTION
**Pull request by teh-cmc**
This PR makes the time series views fully embrace the presence of the cache.

Rather than going through our high-level timepoint-based APIs, these views now go straight to the cache in order to access the raw data, and do their number crunching directly on on that data, allowing them to compute arbitrary many timepoints at a time.
The result is yet another 2-3x performance improvements and arguably much cleaner (read: data-oriented) code: it really is just a matter of splatting ring-buffers on top of each other now (there's still plenty of silly stuff happening of course, we are soooo far from running optimally).

This makes caching a _requirement_ for time series -- the experimental feature flag has no effect anymore, it gotta go.
That's where I wanted to go in the first place, but I didn't think we'd do it in this release.

At this point, I cannot think of any good reason to keep the uncached path (`MaybeCachedComponentData` and the associated experimental feature flags), but I can think of plenty reasons to get rid of it (_cough cough_ #5015).
The next PR will remove the uncached path.

(The commit log is a mess of past experiments, your best bet is to ignore it)

---

#### Single series single-thread benchmark (100k points).

`0.12.1`:
image

`main`:
image

now:
image

#### Same thing, but 1M points

`0.12.1`:
image

`main`:
image

now:
image

---

- Fixes #5017 

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/5048/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/5048/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/5048/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/5048)
- [Docs preview](https://rerun.io/preview/7a3a5c15ea1c14639a081edcb747dee2338d1dc7/docs) 
- [Examples preview](https://rerun.io/preview/7a3a5c15ea1c14639a081edcb747dee2338d1dc7/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_space_view_time_series/src/legacy_visualizer_system.rs [hunk 1]
```diff
@@ -152,80 +153,92 @@ impl LegacyTimeSeriesSystem {
...
-                            MaybeCachedComponentData::iter_or_repeat_opt(&labels, scalars.len()),
-                        ) {
-                            let mut point = PlotPoint {
-                                time: time.as_i64(),
-                                value: scalar.0,
-                                ..default_point.clone()
-                            };
-
-                            // Make it as clear as possible to the optimizer that some parameters
-                            // go completely unused as soon as overrides have been defined.
```
Comment (teh-cmc) [inline crates/re_space_view_time_series/src/legacy_visualizer_system.rs:193]: still a valuable comment for whoever comes next

File: crates/re_query_cache/src/cache.rs [hunk 2]
```diff
@@ -797,6 +797,16 @@ impl CacheBucket {
...
     }
 
+    /// Get the raw batches for the specified non-optional component.
+    #[inline]
+    pub fn component<C: Component + Send + Sync + 'static>(&self) -> Option<&FlatVecDeque<C>> {
+        let data = self
+            .components
+            .get(&C::name())
+            .and_then(|data| data.as_any().downcast_ref::<FlatVecDeque<C>>())?;
+        Some(data)
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:807]:
```suggestion
        self
            .components
            .get(&C::name())?
            .as_any().downcast_ref::<FlatVecDeque<C>>()
```

File: crates/re_query_cache/src/cache.rs [hunk 3]
```diff
@@ -810,6 +820,18 @@ impl CacheBucket {
...
+    /// Get the raw batches for the specified optional component.
+    #[inline]
+    pub fn component_opt<C: Component + Send + Sync + 'static>(
+        &self,
+    ) -> Option<&FlatVecDeque<Option<C>>> {
+        let data = self
+            .components
+            .get(&C::name())
+            .and_then(|data| data.as_any().downcast_ref::<FlatVecDeque<Option<C>>>())?;
+        Some(data)
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:832]:
```suggestion
        self
            .components
            .get(&C::name())?
            .data.as_any().downcast_ref::<FlatVecDeque<Option<C>>>()
```

Comment (emilk) [reply crates/re_query_cache/src/cache.rs:832]: Probably `downcast_ref()` works as well

File: crates/re_space_view_time_series/src/legacy_visualizer_system.rs [hunk 4]
```diff
@@ -152,80 +153,92 @@ impl LegacyTimeSeriesSystem {
...
+                        // Allocate all points.
+                        points = times.map(|time| PlotPoint {
+                            time,
+                            ..default_point.clone()
+                        }).collect_vec();
+
+                        // Fill in values.
+                        for (i, scalar) in scalars.range(entry_range.clone()).enumerate() {
+                            if scalar.len() > 1 {
+                                re_log::warn_once!("found a scalar batch -- those have no effect");
```
Comment (emilk) [inline crates/re_space_view_time_series/src/legacy_visualizer_system.rs:171]: can we print the entity path to it?

File: crates/re_space_view_time_series/src/legacy_visualizer_system.rs [hunk 5]
```diff
@@ -152,80 +153,92 @@ impl LegacyTimeSeriesSystem {
...
-                                        // Common-case optimization
-                                        re_renderer::Color32::from_rgb(r, g, b)
-                                    } else {
-                                        re_renderer::Color32::from_rgba_unmultiplied(r, g, b, a)
+                        // Fill in series kind -- if available _and_ not overridden.
+                        if override_scattered.is_none() {
+                            if let Some(scatterings) = scatterings {
+                                for (i, scattered) in scatterings.range(entry_range.clone()).enumerate() {
+                                    if i >= points.len() {
+                                        re_log::debug_once!("more scatterings attributes than points -- that's a bug");
```
Comment (emilk) [inline crates/re_space_view_time_series/src/legacy_visualizer_system.rs:181]: a bug in what - the data or the code? If so, what part of the code? Can we print something actionable here, like the entity path?

Comment (emilk) [reply crates/re_space_view_time_series/src/legacy_visualizer_system.rs:181]: here and below

File: crates/re_space_view_time_series/src/line_visualizer_system.rs [hunk 6]
```diff
@@ -140,51 +141,46 @@ impl SeriesLineSystem {
...
+                        // Allocate all points.
+                        points = times.map(|time| PlotPoint {
+                            time,
+                            ..default_point.clone()
+                        }).collect_vec();
+
+                        // Fill in values.
+                        for (i, scalar) in scalars.range(entry_range.clone()).enumerate() {
+                            if scalar.len() > 1 {
+                                re_log::warn_once!("found a scalar batch -- those have no effect");
```
Comment (emilk) [inline crates/re_space_view_time_series/src/line_visualizer_system.rs:170]: again

**Review by emilk — APPROVED** [review]
Looks good, but please try to make the log messages actionable

**Pull request merged**